### PR TITLE
Reactant `LandModel` without vegetation

### PR DESCRIPTION
**Pull request by maximilian-gelbrecht**
This ensures Reactant compatibility of a `LandModel` without vegetation.

### CODE REVIEW

File: src/processes/soil/hydrology/soil_hydraulic_closures.jl [hunk 1]
```diff
@@ -116,8 +116,10 @@ end
     por = porosity(i, j, k, grid, fields, strat, bgc)
     sat_res = residual_saturation(get_hydraulic_properties(hydrology))
     # compute matric pressure head
-    ψm = inv_swrc(max(sat * por, sat_res); θsat = por)
-    @assert isfinite(ψm) "NaN/infinite matric potential ψm = $ψm with sat=$sat por=$por resid=$sat_res"
+    # Note that `sat` is clamped from above as well as below: the inverse SWRC is evaluated with a
+    # non-negative base (see `VanGenuchten`/`BrooksCorey`), which oversaturated states (sat > 1) would
+    # violate, raising a `DomainError` from `^` — a throw path that cannot be compiled in a kernel.
+    ψm = inv_swrc(clamp(sat * por, sat_res, por); θsat = por)
```
Comment (maximilian-gelbrecht) [inline src/processes/soil/hydrology/soil_hydraulic_closures.jl:119]: @bgroenks96 I am a bit confused by this one. Claude did this automatically, do you think this is okay?

Comment (maximilian-gelbrecht) [reply src/processes/soil/hydrology/soil_hydraulic_closures.jl:119]: The `@assert` definitely has to be dropped though.

Comment (bgroenks96) [reply src/processes/soil/hydrology/soil_hydraulic_closures.jl:119]: It's fine, but remove or condense the verbose comment.

File: src/processes/soil/hydrology/soil_hydraulic_properties.jl [hunk 2]
```diff
@@ -229,8 +229,11 @@ function hydraulic_conductivity(
         Ω = hydraulics.unsat_hydraulic_cond.impedance # scaling parameter for ice impedance
         I_ice = 10^(-Ω * (1 - f)) # ice impedance factor
         K_sat = saturated_hydraulic_conductivity(hydraulics, soil.solid.texture)
-        # We use `complex` types here to permit illegal state values which may occur when using adaptive time steppers.
-        K = abs(K_sat * I_ice * sqrt(complex(θw / θsat)) * (1 - complex(1 - complex(θw / θsat)^(n / (n + 1)))^((n - 1) / n))^2)
+        # Effective saturation, clamped to the unit interval so that every fractional power below is
+        # evaluated at a non-negative base. Illegal state values (θw > θsat or θw < 0), which may occur
+        # when using adaptive time steppers.
+        S = clamp(θw / θsat, zero(θw), one(θw))
+        K = K_sat * I_ice * sqrt(S) * (one(S) - (one(S) - S^(n / (n + 1)))^((n - 1) / n))^2
```
Comment (bgroenks96) [inline src/processes/soil/hydrology/soil_hydraulic_properties.jl:240]: Also fine.

File: src/processes/surface/skin_temperature.jl [hunk 3]
```diff
@@ -80,7 +80,7 @@ Seed the prognostic `skin_temperature` with the current `ground_temperature` so
 nonlinear solve starts from a physically sensible guess close to the root.
 """
 function initialize!(state, grid, ::ImplicitSkinTemperature, args...)
-    set!(state.skin_temperature, state.ground_temperature)
+    interior(state.skin_temperature) .= interior(state.ground_temperature)   # Now set! for better Reactant compat, set! might make a CPU detour
```
Comment (bgroenks96) [inline src/processes/surface/skin_temperature.jl]: The comment doesn't match the change here. Why do we need to stop using `set!`? I would rather keep that.

**Review by bgroenks96 — DISMISSED** [review]
Looks overall good, just look at the `set!` thing. That would be problematic if Reactant can't handle that.

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: It's just a typo "No set!". 

The problem is that `set!` might make CPU detours and the interpolation isn't fully Reactant compatible.

File: src/processes/soil/hydrology/soil_hydraulic_closures.jl (hunk 1 above)
Comment (maximilian-gelbrecht) [reply src/processes/soil/hydrology/soil_hydraulic_closures.jl:119]: Removed

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: Did you create an Oceananigans issue about that?

I don't see any reason why `set!` should make CPU detours when setting a `Field` to another `Field`.

Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: No, not yet

Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: Ok, then please do that and link it here, and add a TODO comment in the code.

**Comment by maximilian-gelbrecht** [comment]
Oh I thought this would already pass but it turns out there is another upstream Oceananigans Reactant issue with boundary conditions that we hit here. I converted this to draft for now. 

The main problem is the 2D indexing of `Fields` as it seems.  

I need to sit down and look at this properly to understand exactly what we need and maybe fix the issue upstream in Oceananigans.

File: src/solvers/newton.jl [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+straight-line code.
+
+Unlike [`RootSolver`](@ref) and [`FixedPointSolver`](@ref) there is no convergence test and hence no
+data-dependent loop bound. That matters in two places:
+
+- **Reactant**: a convergence-tested loop lowers to an `scf.while` with a dynamic trip count, which the
+  StableHLO raise pass cannot lift. An unrolled loop raises cleanly.
+- **Reverse-mode AD**: a fixed trip count avoids a dynamically-sized tape.
+"""
+struct NewtonSolver{NF, iterations} end
```
Comment (bgroenks96) [inline src/solvers/newton.jl:17]: Why? We already had a Newton solver before and got rid of it in favor of RootSolvers.jl.

Can't you just disable the convergence test for the RootSolvers one?

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: Also, this is again something that should be fixed with a custom rule... the AD system should not be differentiating through the iterations.

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: No custom rules with Reactant, but you are probably right with RootSolvers.

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: Is there any documentation or roadmap anywhere about custom rules? I can't find anything, and I find this really concerning.

**Comment by bgroenks96** [comment]
>  The main problem is the 2D indexing of Fields as it seems.

It would not be so objectionable to do `field[i, j, 1]` rather than `field[i, j]` everywhere. We already have to do this anyway when using `setindex!`.

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: This happens apparently because `skin_temperature` is `Field{Center, Center, Nothing}` and `ground_temperature` is `Field{Center, Center, Center}`. This triggers an interpolation then.

**Comment by maximilian-gelbrecht** [comment]
> > The main problem is the 2D indexing of Fields as it seems.
> 
> It would not be so objectionable to do `field[i, j, 1]` rather than `field[i, j]` everywhere. We already have to do this anyway when using `setindex!`.

The culprit is in Oceananigans though https://github.com/CliMA/Oceananigans.jl/issues/5846

File: src/solvers/newton.jl (hunk 4 above)
Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: I am not terribly concerned around it as even the worst case of just differentiating through the maximum itertation number might not be that bad. But we are also not alone with this https://github.com/CliMA/Oceananigans.jl/issues/5844

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: Ahh... I guess this would also happen if `skin_temperature` were defined at a `Face` (as it could be). 

The interpolation is probably pointless in this case unless the soil temperature has a Dirichlet boundary condition.

File: src/solvers/newton.jl (hunk 4 above)
Comment (bgroenks96) [reply src/solvers/newton.jl:17]: An AD system without the ability to define gradients is not a complete AD system...

But anyway, I think we get around this by using `ignore_deriatives` which should be supported by Reactant, see https://github.com/EnzymeAD/Reactant.jl/pull/1373

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: RootSolvers isn't really compatible with Reactant as it is, it seems to me. Too many runtime conditions that can't be deactivated. I think I would keep this simple newton solver for the Reactant for now and then we might revisit this later depending on how we proceed with Reactant support (and the rest of the model)

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: I would prefer to just do `ignore_derivatives` if possible?

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: I am not sure how `ignore_derivatives` should help here. This isn't only about taking derivatives. It's about running the model with Reactant also forward.

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: But the behaviour you intended for the initialization is to just copy the values, or do we actually want the interpolation in case of different locations?

File: src/solvers/newton.jl (hunk 4 above)
Comment (bgroenks96) [reply src/solvers/newton.jl:17]: What precisely is the issue? It fails to compile entirely?

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: It's things like this that make me not so convinced by Reactant. There seems to be a growing number of things that we now can't do in our code because of it, and that seems to be effectively just giving up on the original promise of "differentiate arbitrary Julia code". By the way, JAX absolutely can differentiate loops with tracked conditionals, see [jax.lax.while_loop](https://docs.jax.dev/en/latest/_autosummary/jax.lax.while_loop.html).

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: Reactant can also absolutely do that, but then the loop needs to be implemented like that e.g. with a `@trace` but RootSolvers.jl don't do that https://enzymead.github.io/Reactant.jl/stable/tutorials/control-flow#Loops

But it's also true that both for Reactant and for GPU in general these kind of data dependent kernels might just be undesirable performance wise as well because they may cause too many branch divergences.

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: Then let's just make a PR to RootSolvers.jl to add `@trace`. ReactantCore is a lightweight package, and as long as it doesn't affect their code when *not* using Reactant, I don't see why they should object to it.

Comment (olivierbonte) [reply src/solvers/newton.jl:17]: First attempt at RootSolvers.jl compatible with Reactant (not yet complete) https://github.com/CliMA/RootSolvers.jl/pull/99 

The main problem that I am now at is that there is data-dependent while loop see https://github.com/olivierbonte/RootSolvers.jl/blob/db2a8349ecfaf9cae00516dac139286f52e930c8/src/RootSolvers.jl#L1851. As in the MWE found at https://github.com/CliMA/Oceananigans.jl/issues/5844, it is also here the case you you can make it work for the normal forward evaluation, but the AD fails. Following the suggestion of https://github.com/CliMA/Oceananigans.jl/issues/5844#issuecomment-5171628126 I could try to rewrite this.

So unfortunately not super straightforward to make it compatible. A rule system for AD would be nice, unfortunate that this is not (yet?) in Reactant.

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: That's already a great start Olivier! 

In the long run it would be nice if RootSolvers would be compatible, best with adjoint rules, but for the time being I am really not too worried. The Reactant model is already very very fast without any attempt to profile it (normal runs and AD), and if we have a separate fixed iteration version for Reactant for the time being that's not so bad and still quite performant.

**Comment by maximilian-gelbrecht** [comment]
> > > The main problem is the 2D indexing of Fields as it seems.
> > 
> > 
> > It would not be so objectionable to do `field[i, j, 1]` rather than `field[i, j]` everywhere. We already have to do this anyway when using `setindex!`.
> 
> The culprit is in Oceananigans though [CliMA/Oceananigans.jl#5846](https://github.com/CliMA/Oceananigans.jl/issues/5846)

This issue is resolved (but not yet released). 

I'll pick up this PR when I find some time, maybe on the way home from JuliaCon.

File: src/solvers/newton.jl (hunk 4 above)
Comment (bgroenks96) [reply src/solvers/newton.jl:17]: @olivierbonte I also attempted it (not sure if I pushed it to my fork) and ran into the same issues. I gave up after a couple of hours. I don't think it's entirely an issue of RootSolvers.jl being incompatible; in theory Reactant can handle data-dependent loops. But I hit a point where it looked like bugs in Reactant/StableHLO and just gave up.

Comment (olivierbonte) [reply src/solvers/newton.jl:17]: Alright, fixed point iteration for Reactant for now sounds reasonable!
Good to know you also gave it a shot @bgroenks96! I'll try to look at it a bit more if I find the time, but not sure If I will get the AD working. Do you think there is value in trying to get a PR merged in RootSolvers.jl that works for forward evaluations (but not yet for Enzyme)? Because that I think should work

Comment (bgroenks96) [reply src/solvers/newton.jl:17]: If the changes to the code are not too substantial, I would try it. But the CliMA project is pretty explicitly disinterested in AD so I doubt they will accept a major rewrite for Reactant.

Comment (maximilian-gelbrecht) [reply src/solvers/newton.jl:17]: If you can reduce it to a Reactant MWE if something isn't working that would be great as well though, we should absolutely post it in their repo. Usually responses are quick.

Comment (olivierbonte) [reply src/solvers/newton.jl:17]: https://github.com/CliMA/RootSolvers.jl/pull/99 now works for primal evaluations. Enzyme autodiff still WIP

File: src/utils/kernel_utils.jl [hunk 5]
```diff
@@ -1,3 +1,18 @@
+"""
+    @assert_kernel cond [text]
```
Comment (maximilian-gelbrecht) [inline src/utils/kernel_utils.jl:18]: I added `@assert_kernel` macro now that just disappears within a Reactant compile context, but otherwise just forwards to `Base.@assert`. This way we can still assert basically as before.

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: I intended for it to just be copied.

**Comment by maximilian-gelbrecht** [comment]
This works now. I would keep the fixed iteration `NewtonSolver` as a temporary solution for Reactant until we have something different figured out (like RootSolvers.jl compatability).

**Comment by olivierbonte** [comment]
@maximilian-gelbrecht have you tried the RootSolvers.jl version at https://github.com/CliMA/RootSolvers.jl/pull/99? This was working okay for normal (non AD) reactant code evaluations but haven't tried it yet within Terrarium.

File: src/processes/surface/skin_temperature.jl (hunk 3 above)
Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: Also, I still don't understand why interpolation would happen on CPU. That seems like a bug? @glwagner

Comment (glwagner) [reply src/processes/surface/skin_temperature.jl]: Hmm, it's true that different locations will trigger interpolation I think...

https://github.com/CliMA/Oceananigans.jl/blob/c8a95be84673affcefe59aa12592de7f6f88f60c/src/Fields/set!.jl#L164

and in particular here

https://github.com/CliMA/Oceananigans.jl/blob/c8a95be84673affcefe59aa12592de7f6f88f60c/src/Fields/set!.jl#L182-L183

we could perhaps fix this edge case by changing the condition `ℓu == ℓv` to a more specific condition that fails only for `Center` and `Face` but returns `true` if one of the locations is `Nothing`. I think that would allow this case to work right?

Something we were trying to prevent (which I believe did occur as a bug) is merely checking the size of the fields. As an example, `set!(u, v)` definitely should not behave like `interior(u) .= interior(v)` if `u` and `v` are at `Center` and `Face` in a periodic direction.

Unrelated, it also looks like there's a bug if the two fields are on the same size grid when those grids cover different spatial extents.

Also, we really do need Reactant support for interpolation. I think this is kind of important.

I don't believe that anything should go to CPU though. That should only happen if the target is on CPU.

Comment (bgroenks96) [reply src/processes/surface/skin_temperature.jl]: > we could perhaps fix this edge case by changing the condition ℓu == ℓv to a more specific condition that fails only for Center and Face but returns true if one of the locations is Nothing. I think that would allow this case to work right?

Doesn't it already do that with `reduced_singleton_dimension`? What is `N` here? The length of the dimension?

Comment (glwagner) [reply src/processes/surface/skin_temperature.jl]: Indeed, it should have. Except that reactant defines its own `set_to_field!`

So the somewhat funny situation is that the Reactant extension of `set_to_field!` actually causes `set!` to fail with Reactant, when it would otherwise succeed.

It's fixed in a new PR.

Comment (glwagner) [reply src/processes/surface/skin_temperature.jl]: Also on this:

> I don't believe that anything should go to CPU though. That should only happen if the target is on CPU.

we hit CPU only for interpolation (which isn't yet supported with Reactant), which is what we fall back to if we decide that broadcasting / copying will not work.

Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: Thanks Greg! 

I reverted to `set!` and it's not yet working here end-to-end, I'll investigate a bit later.

Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: Ah sorry, stupid me, I only added the branch in the main and normal test env of Terrarium and not in the Reactant test env. It does actually work.

Comment (maximilian-gelbrecht) [reply src/processes/surface/skin_temperature.jl]: Ah sorry, stupid me, I only added the branch in the main and normal test env of Terrarium and not in the Reactant test env. It does actually work.

**Comment by maximilian-gelbrecht** [comment]
Further Oceananigans issues that blocked the vegetation model are addressed in https://github.com/CliMA/Oceananigans.jl/pull/5880

File: src/solvers/newton.jl (hunk 4 above)
Comment (bgroenks96) [reply src/solvers/newton.jl:17]: I guess we will keep the Newton solver for now.

**Comment by bgroenks96** [comment]
@maximilian-gelbrecht Is this ready to be merged?

**Comment by maximilian-gelbrecht** [comment]
Yes with the new Oceananigans version released yesterday, we can merge this.

**Review by bgroenks96 — APPROVED** [review]
